src/datachain/catalog/catalog.py

```diff
@@ -1506,13 +1550,50 @@ def query(
         if capture_output:
             popen_kwargs = {"stdout": subprocess.PIPE, "stderr": subprocess.STDOUT}
 
-        with subprocess.Popen(cmd, env=env, **popen_kwargs) as proc:  # noqa: S603
-            if capture_output:
-                args = (proc.stdout, output_hook)
-                thread = Thread(target=_process_stream, args=args, daemon=True)
-                thread.start()
-                thread.join()  # wait for the reader thread
+        def signal_handler(sig: int, frame: Any) -> NoReturn:
+            raise TerminationSignal(sig)
+
+        orig_handler = signal.getsignal(signal.SIGTERM)
+        signal.signal(signal.SIGTERM, signal_handler)
 
+        thread = None
+        with subprocess.Popen(cmd, env=env, **popen_kwargs) as proc:  # noqa: S603
+            logger.info("Running script with PID %s", proc.pid)
+            try:
+                if capture_output:
+                    args = (proc.stdout, output_hook)
+                    thread = Thread(target=_process_stream, args=args, daemon=True)
+                    thread.start()
+
+                proc.wait()
+            except (KeyboardInterrupt, TerminationSignal) as exc:
+                if orig_handler is not None:
+                    signal.signal(signal.SIGTERM, orig_handler)
+
+                logging.info("Shutting down process %s, received %r", proc.pid, exc)
+
+                # If process is running in the foreground, TTY sends signals to all
+                # processes in the process group. So, by default, we don't send
+                # interrupt signal to the child process again.
+                # If interrupt needs to be sent, this can be enabled by setting
+                # send_interrupt to True.
+                shutdown_process(
+                    proc,
+                    send_interrupt,
+                    interrupt_timeout=interrupt_timeout,
+                    terminate_timeout=terminate_timeout,
+                )
+                if proc.returncode:
+                    raise QueryScriptCancelError(
+                        "Query script was canceled by user", return_code=proc.returncode
+                    ) from exc
+            finally:
+                if orig_handler is not None:
```

Comment:
it looks like we are triggering it twice?

Reply:
I am trying to reinstate old handler as soon as possible. It's okay if it gets called twice - there can only be one signal handler at a time for a given signal.

Anyway, I need to double-check all the logic here, and also think of what happens when someone resends SIGTERM/SIGINT.

Reply:
We need this. We need to rollback the signal handlers before we try to gracefully shutdown. Otherwise, if users send `SIGTERM` and `SIGINT`, we won't be able to forcefully exit.